pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -224,6 +151,269 @@ func (a *AmbientIndex) matchesScope(scope model.WaypointScope, w *model.Workload
 	return true
 }
 
+func PickBestPod(pods []*v1.Pod) *v1.Pod {
+	var best *v1.Pod
+	for _, p := range pods {
+		if !IsPodRunning(p) {
+			continue
+		}
+		if best == nil || p.CreationTimestamp.After(best.CreationTimestamp.Time) {
+			best = p
+		}
+	}
+	return best
+}
+
+func (a *AmbientIndex) Recompute(name types.NamespacedName) *model.WorkloadInfo {
+	ip := name.Name
+	log := log.WithLabels("controller", "ambient", "ip", ip)
+	log.Debugf("reconcile")
+
+	// Should be 0 or 1 pod, unless we are in a weird state
+	pod := PickBestPod(a.podIPIndex.Lookup(ip))
+
+	// Should be 0 or 1 entry, unless we are in a weird state
+	var entry *networking.WorkloadEntry
+	var configEntry *config.Config
+	if a.workloadEntriesIPIndex != nil {
+		entries := a.workloadEntriesIPIndex.Lookup(ip)
+		if len(entries) > 0 {
+			configEntry = entries[0]
+			entry = serviceentry.ConvertWorkloadEntry(*entries[0])
+		}
+	}
+
+	// We can have many slices (part of many services). Note each slice contains other IPs as well; we need to filter
+	epSlices := a.sliceIPIndex.Lookup(ip)
+	slices.SortFunc(epSlices, func(a, b *discoveryv1.EndpointSlice) bool {
+		// Newer slices win, could be an address moving from 1 to another
+		return a.CreationTimestamp.After(b.CreationTimestamp.Time)
+	})
+
+	if len(epSlices) == 0 && pod == nil && entry == nil {
```

Comment:
it might be nice to abstract the pod/WE into a new Workload type

Reply:
I completely agree, its not super easy to do it well in the current architecture though; but I think long term all the controllers should work in this way. Technically this function _is_ building a Workload type from Pod and WE, though :slightly_smiling_face: